core-bundle/src/Image/Preview/PreviewFactory.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Image\Preview;
+
+use Contao\CoreBundle\Image\ImageFactoryInterface;
+use Contao\CoreBundle\Image\PictureFactory;
+use Contao\CoreBundle\Image\PictureFactoryInterface;
+use Contao\CoreBundle\Image\Studio\Figure;
+use Contao\CoreBundle\Image\Studio\FigureBuilder;
+use Contao\CoreBundle\Image\Studio\Studio;
+use Contao\Image\ImageInterface;
+use Contao\Image\PictureConfiguration;
+use Contao\Image\PictureInterface;
+use Contao\Image\ResizeConfiguration;
+use Contao\Image\ResizeOptions;
+use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\PathUtil\Path;
+
+class PreviewFactory
+{
+    /**
+     * @var iterable<int,PreviewProviderInterface>
+     */
+    private iterable $previewProviders;
+
+    private ImageFactoryInterface $imageFactory;
+    private PictureFactoryInterface $pictureFactory;
+    private Studio $imageStudio;
+    private string $cacheDir;
+    private array $validImageExtensions;
+
+    /**
+     * @param iterable<int,PreviewProviderInterface> $previewProviders
+     */
+    public function __construct(iterable $previewProviders, ImageFactoryInterface $imageFactory, PictureFactoryInterface $pictureFactory, Studio $imageStudio, string $cacheDir, array $validImageExtensions)
+    {
+        $this->previewProviders = $previewProviders;
+        $this->imageFactory = $imageFactory;
+        $this->pictureFactory = $pictureFactory;
+        $this->imageStudio = $imageStudio;
+        $this->cacheDir = $cacheDir;
+        $this->validImageExtensions = $validImageExtensions;
+    }
+
+    public function createPreviewFile(string $path, int $size = 0): string
+    {
+        // Supported image formats do not need an extra preview image
+        if (\in_array(strtolower(pathinfo($path, PATHINFO_EXTENSION)), $this->validImageExtensions, true)) {
+            return $path;
+        }
+
+        // Round up to the next highest power of two
+        $size = (int) (2 ** log($size, 2));
+
+        // Minimum size for previews is 512
+        $size = max(512, $size);
+
+        $cachePath = $this->createCachePath($path, $size);
+        $globPattern = preg_replace('/[*?[{\\\\]/', '\\\\$0', $this->cacheDir.'/'.$cachePath).'.*';
+
+        foreach (glob($globPattern) as $cacheFile) {
+            if (\in_array(pathinfo($cacheFile, PATHINFO_EXTENSION), $this->validImageExtensions, true)) {
+                return $cacheFile;
+            }
+        }
+
+        $first1024Bytes = file_get_contents($path, false, null, 0, 1024);
```

Comment:
Ah, here are the bytes. Again, maybe there's a provider that needs 2 KB to determine support. We should fetch that from them.

Reply:
1cd029830d63475ec9d74eca2c19945e0852ffd2